PyEMD/tests/test_ceemdan.py

```diff
@@ -177,6 +177,15 @@ def test_ceemdan_notParallel(self):
         self.assertTrue(cIMFs.shape[0]>1)
         self.assertTrue(cIMFs.shape[1]==S.size)
 
+    def test_imfs_and_residue_accessor(self):
```

Comment:
Since you added the exception `ValueError` it would be also nice to test that it works as expected. Could you please add a test case when you don't run CEEMDAN and you check that `ValueError`?

Same goes for EEMD and EMD that are updated in the same pull request.

Reply:
Hi @laszukdawid 

I included the tests and thanks to that I found that this exception was not working properly.

For **emd.py** it was working because the IMFs and residue were both declared inside the constructor (`self.__init__`), while in the other EMD variants, these variables were declared only inside another method (like in `def eemd(self, S, T=None, max_imf=-1):`.

Because of that, the conditional was checking for None values in a non-existent variable, the error that actually raises is `AttributeError` instead of `ValueError`.

I see 2 workarounds for this:

- For all Python modules (emd, eemd, ceemdan), transfer the IMF and residue declarations to the constructor, initializing them with None, and then leave the `get_imfs_and_residue` method as it is. So every module becomes consistent.

-  Or bring the IMF and residue declarations in **emd.py** inside the `emd` method (so it is consistent with the other modules) and  then, replace the if statement in `get_imfs_and_residue` method by a try...except statement raising an exception for the `AttributeError`. In the exception, we could print a message and raise it again, or automatically run the method to compute the IMFs and residue using some default input parameters.

I don't know which one you prefer. In my opinion, running the method automatically after the exception is a bad idea because some users may use the code blindly without worrying about the input parameters. In this commit, the solution that I am adopting is the first one, but let me know if you want me to change it.

Reply:
Agree on the solution :) Good instinct!